Safe Z homing for Power Loss Recovery

SwiftNick opened the pull request:
### Requirements
Add option to home the Z axis during power loss recovery 
During recovery, all axes are homed, then Z position is restored before X & Y   

### Description
Adds a parameters to the POWER_LOSS_RECOVERY block in configuration_adv.h
If enabled, during PrintJobRecovery::resume() will now 
1) raise Z by POWER_LOSS_ZRAISE
2) home X & Y
3) move to POWER_LOSS_ZHOME_XPOS and POWER_LOSS_ZHOME_YPOS
4) home Z
5) raise Z to saved Z position + POWER_LOSS_ZRAISE
6) continue with standard print resume
 
The feature is disabled by default and will be blocked if Z_SAFE_HOMING is enabled.

### Benefits
This allows print to resume after power loss if the Z position cannot be guaranteed provided that  there is a safe region to home the Z axis

### Modified files
configuration_adv.h
feature\power_loss_recovery.cpp
inc\SanityCheck.h

SwiftNick committed:
Add switch to restore lost Z-axis position during power loss recovery

If printer loses the Z-axis position during power loss, it needs to be recovered safely during the power loss recovery stage.
1) Safely home the Z axis (assumes print can be avoided)
2) Restore the Z position before XY

thinkyhead commented:
- If the print head has fallen down due to a power-loss, the print is not recoverable.
- We don't yet support the Anycubic Chiron.

SwiftNick commented:
The head drops about 2mm due to the pitch of the lead screw, but the print can be recovered, as at the point the power goes, the current z position is saved to the PLR file.
If Z is homed after power loss then the position is recovered just fine.
The machine is designed for this and Anycubic's custom firmware using v1.3.0 supports it too.

Regarding general support for the Chiron, I have written a new pin def, config file and an interface for the factory fitted TFT panel using the extensible UI API which together provides full support for the factory build Chiron.

Need to have a chat about how to do a PR for that, as it has dependencies for config files and core Marlin which are now 2 separate repositories.

Any suggestions welcome!

tpruvot commented:
G28 Z might collide the half printed part...

SwiftNick commented:
Yes that is possible, 
In the owners guide for the Chiron it states that the front 10cm of the bed MUST be avoided if you intend to use power loss recovery, otherwise you will hit the printed object.
This is how they have implemented it in their custom version of Marlin.

TBH it would have been better if they used a Z Axis lead screw with a shallower pitch , so the axis doesn't move when powered down.

thinkyhead commented:
I have a Chiron here, so I can assist with testing. I replaced its board before finding out that a special driver is needed before you can see it over USB, so I have to re-check my wires and maybe put back the original board because I did a sloppy job.

studiodyne commented:
YES ITS RECOVERABLE , but with my idea ,  to add a ZMAX endstop instead of a ZMIN , the homing will low down the bed to find the endstop and will raise to the perfect z value 

But , need to install a ZMAX endstop 
All machine with bed that go down on shutdown must have a reverse homing , with the endstop on zmax , if not , recovery will ever be impossible

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bf2_plr_update_16909

thinkyhead committed:
Fixup

thinkyhead committed:
Update powerloss.cpp

thinkyhead committed:
Update powerloss.cpp

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bf2_plr_update_16909

thinkyhead commented on Marlin/src/feature/powerloss.cpp:
```diff
@@ -407,23 +407,23 @@ void PrintJobRecovery::resume() {
+    dtostrf(info.current_position.y, 1, 3, str_2)
+  );
+  gcode.process_subcommands_now(cmd);
+  
```
This is a bad idea. You don't want to do a sideways move when Z is at the lower height. Power Loss Recovery currently does the sideways move before lowering the Z axis in order to avoid knocking over the model.

SwiftNick replied:
I had swapped the Z and XY resume moves, but a better option would be to restore Z above the print before moving back to resume printing.

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into bf2_plr_update_16909

8Moshe8 commented:
I think homing Z during power loss recovery as an option could be a great feature.
A good algorithm will be:
- raising Z 
- homing XY
- moving to a desired XY position (as specified by user in configuration_adv)
- then homing Z
- raising Z to the correct height
- Resuming printing.
An optional purge filament/continue sub menu like in filament change would help.

if the feature will be realized in the a similar way and users could disable it or customize it for their machines it would be great and a solid addition
I would greatly appreciate if it will be added in the future to Marlin 2.0

SwiftNick committed:
After home XY, Restore Z above print

Standard resume now: 
1) raises Z by POWER_LOSS_ZRAISE
Added:
2) home XY
3) home Z
4) Raise Z to savedZ + POWER_LOSS_ZRAISE
Standard resume then:
5) restores XY
6) drops Z back to savedZ
7) resumes print.

8Moshe8 commented:
Previously (Marlin 2.0.5.3) it didn`t home Z at all during recovery.
it raised Z. Then homed xy. Then recovered xy. Then "recovered z" (assumed
z stayed as it was) then continued printing.

The best solution for a printer like ender 3 that has a safe place on xy
plane to home Z without it knocking down print is to:
raise Z, Then home xy, Move to the safe place outside the bed, home Z,
raise Z to previous Z+some delta, restore xy, restore Z and continue
printing.
for example for an ender 3 the safe space is at Xmin Ymin. so you will do a
normal home immediately after moving up (won`t work with safe homing option
or probes),
then raising to Z+delta (without moving in xy plane) restoring xy restoring
z and continuing.
this "safe spot" for homing is necessary for POWER_LOSS_ZHOME to work as
expected.

if you home z before safely moving to the safe place at the origin you may
collide with the print or build plate.
it is important to home xy before moving to safe space and homing z.

I highly appreciate your contribution and attention. Thank you very much
for your work. I would test the new feature but would also love
if you consider my suggestion.

Moshe.

SwiftNick committed:
Add settings for power loss Z home safe position

SwiftNick committed:
Move to safe Z Home position on power loss recovery

Settings added to specify a safe XY position to use for Z homing after power loss that avoids colliding with print.

SwiftNick commented:
I've added some options to the POWER_LOSS_ZHOME switch in Advanced Config so you can set a 'safe' position to restore z after power loss.
POWER_LOSS_ZHOME_XPOS and POWER_LOSS_ZHOME_YPOS
I need to look at the sanity check though as this may not work with Z_SAFE_HOMING

SwiftNick committed:
Added test for POWER_LOSS_ZHOME 

Cannot be used when Z_SAFE_HOMING enabled

8Moshe8 commented:
I have modified the original powerloss.cpp file using SwiftNick`s code to fix the bugs that prevented hot end temperature from restoring.
I compiled the code and tested the feature on my ender 3 and it worked as expected like a charm.
How can I submit my changes?
I am new to GitHub and this will be the first time I can contribute something.

SwiftNick made changes to SanityCheck.h and configuration_adv.h
both those files are required.
I changed only powerloss.cpp to fix the issue above.

SaschaKP commented on Marlin/src/inc/SanityCheck.h:
```diff
@@ -2570,6 +2570,12 @@ static_assert(   _ARR_TEST(3,0) && _ARR_TEST(3,1) && _ARR_TEST(3,2)
 #endif
 
+#if ENABLED(POWER_LOSS_ZHOME)
+  #if ENABLED(Z_SAFE_HOMING)
```
for users who have a ZMAX endstop in addition to the ZMIN endstop it should be made an option to use that switch, but as I can see there is no way to ZHOME on opposite direction, when the switch is declared on the other side, so this could require some work. Also, that addition, could make it work on printers with Z_SAFE_HOMING

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/16909

thinkyhead committed:
Followup to merge

thinkyhead commented:
Brought this up to date with the latest settings and PLR behavior.

I added a `G28 Z U` with the plan to make "`U`" a new flag in `G28` which will home the Z axis "unsafely" because `G28 Z` by itself will move XY to the "safe homing position" and that might not be what you want. But on the other hand, if you _are_ using a probe and Z safe homing, you _really_ should be avoiding this new option, so maybe it makes more sense to just forbid the new option when using Z safe homing.

thinkyhead commented:
Looks like that's covered already, so one more quick patch…

thinkyhead committed:
Merge remote-tracking branch 'upstream/bugfix-2.0.x' into pr/16909

thinkyhead committed:
sanity check for POWER_LOSS_ZHOME_POS

thinkyhead committed:
test

thinkyhead committed:
Fix compile warning
